File: C++/source/Framework/Utils/WaveSynchronousLinearSequencerContext.cpp
/***************************************************************************
 *   Copyright (C) 2005 Vidyasagara Guntaka                                *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Utils/WaveSynchronousLinearSequencerContext.h"
#include "Framework/Messaging/Local/WaveMessage.h"
#include "Framework/Utils/AssertUtils.h"
#include "Framework/ObjectModel/WaveManagedObject.h"

#include <pthread.h>

#include <time.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>

using namespace std;

namespace WaveNs
{

WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext (WaveMessage *pWaveMessage, WaveElement *pWaveElement, WaveSynchronousLinearSequencerStep *pSteps, UI32 numberOfSteps)
{
    m_pWaveMessage                 = pWaveMessage;
    m_pWaveAsynchronousContext     = NULL;
    m_pWaveElement                 = pWaveElement;
    m_pSteps                        = NULL;
    m_numberOfSteps                 = numberOfSteps;
    m_currentStep                   = 0;
    m_isTransactionStartedByMe      = false;
    m_isADelayedCommitTransaction   = false;
    m_completionStatus              = WAVE_MESSAGE_ERROR;

    m_clockId                   = 0;

    if (3 > m_numberOfSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext : There should be atleast three steps to run a Wave Linear Sequencer. " << numberOfSteps << " Steps were specified." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return;
    }

    m_pSteps = new WaveSynchronousLinearSequencerStep[m_numberOfSteps];

    if (NULL == m_pSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext : Could not allocate memory for steps." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return;
    }

    UI32 i = 0;

    for (i = 0; i < m_numberOfSteps; i++)
    {
        m_pSteps[i] = pSteps[i];
    }

    return;
}

WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext (WaveAsynchronousContext *pWaveAsynchronousContext, WaveElement *pWaveElement, WaveSynchronousLinearSequencerStep *pSteps, UI32 numberOfSteps)
{
    m_pWaveMessage                 = NULL;
    m_pWaveAsynchronousContext     = pWaveAsynchronousContext;
    m_pWaveElement                 = pWaveElement;
    m_pSteps                        = NULL;
    m_numberOfSteps                 = numberOfSteps;
    m_currentStep                   = 0;
    m_isTransactionStartedByMe      = false;
    m_isADelayedCommitTransaction   = false;
    m_completionStatus              = WAVE_MESSAGE_ERROR;

    m_clockId                       = 0;

    if (3 > m_numberOfSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext : There should be atleast three steps to run a Wave Linear Sequencer. " << numberOfSteps << " Steps were specified." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return;
    }

    m_pSteps = new WaveSynchronousLinearSequencerStep[m_numberOfSteps];

    if (NULL == m_pSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext : Could not allocate memory for steps." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return;
    }

    UI32 i = 0;

    for (i = 0; i < m_numberOfSteps; i++)
    {
        m_pSteps[i] = pSteps[i];
    }

    return;
}

WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext (const WaveSynchronousLinearSequencerContext &prismSynchronousLinearSequencerContext)
{
    m_pWaveMessage                 = prismSynchronousLinearSequencerContext.m_pWaveMessage;
    m_pWaveAsynchronousContext     = prismSynchronousLinearSequencerContext.m_pWaveAsynchronousContext;
    m_pWaveElement                 = prismSynchronousLinearSequencerContext.m_pWaveElement;
    m_pSteps                        = prismSynchronousLinearSequencerContext.m_pSteps;
    m_numberOfSteps                 = prismSynchronousLinearSequencerContext.m_numberOfSteps;
    m_currentStep                   = prismSynchronousLinearSequencerContext.m_currentStep;
    m_isTransactionStartedByMe      = prismSynchronousLinearSequencerContext.m_isTransactionStartedByMe;
    m_isADelayedCommitTransaction   = prismSynchronousLinearSequencerContext.m_isADelayedCommitTransaction;
    m_completionStatus              = WAVE_MESSAGE_ERROR;

    m_clockId                       = 0;

    if (3 > m_numberOfSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext : There should be atleast three steps to run a Wave Linear Sequencer. " << m_numberOfSteps << " Steps were specified." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return;
    }

    m_pSteps = new WaveSynchronousLinearSequencerStep[m_numberOfSteps];

    if (NULL == m_pSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::WaveSynchronousLinearSequencerContext : Could not allocate memory for steps." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return;
    }

    UI32 i = 0;

    for (i = 0; i < m_numberOfSteps; i++)
    {
        m_pSteps[i] = prismSynchronousLinearSequencerContext.m_pSteps[i];
    }

    return;
}

WaveSynchronousLinearSequencerContext::~WaveSynchronousLinearSequencerContext ()
{
    if (NULL != m_pSteps)
    {
        delete[] m_pSteps;
    }

    garbageCollect ();
}

WaveSynchronousLinearSequencerContext &WaveSynchronousLinearSequencerContext::operator = (const WaveSynchronousLinearSequencerContext &prismSynchronousLinearSequencerContext)
{
    m_pWaveMessage                 = prismSynchronousLinearSequencerContext.m_pWaveMessage;
    m_pWaveAsynchronousContext     = prismSynchronousLinearSequencerContext.m_pWaveAsynchronousContext;
    m_pWaveElement                 = prismSynchronousLinearSequencerContext.m_pWaveElement;
    m_pSteps                        = prismSynchronousLinearSequencerContext.m_pSteps;
    m_numberOfSteps                 = prismSynchronousLinearSequencerContext.m_numberOfSteps;
    m_currentStep                   = prismSynchronousLinearSequencerContext.m_currentStep;
    m_isTransactionStartedByMe      = prismSynchronousLinearSequencerContext.m_isTransactionStartedByMe;
    m_isADelayedCommitTransaction   = prismSynchronousLinearSequencerContext.m_isADelayedCommitTransaction;
    m_completionStatus              = prismSynchronousLinearSequencerContext.m_completionStatus;

    m_clockId                       = 0;

    if (3 > m_numberOfSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::operator = : There should be atleast three steps to run a Wave Linear Sequencer. " << m_numberOfSteps << " Steps were specified." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return (*this);
    }

    m_pSteps = new WaveSynchronousLinearSequencerStep[m_numberOfSteps];

    if (NULL == m_pSteps)
    {
        cerr << "WaveSynchronousLinearSequencerContext::operator = : Could not allocate memory for steps." << endl;
        waveAssert (false, __FILE__, __LINE__);
        return (*this);
    }

    UI32 i = 0;

    for (i = 0; i < m_numberOfSteps; i++)
    {
        m_pSteps[i] = prismSynchronousLinearSequencerContext.m_pSteps[i];
    }

    return (*this);
}

void WaveSynchronousLinearSequencerContext::advanceCurrentStep ()
{
    m_currentStep++;
}

ResourceId WaveSynchronousLinearSequencerContext::executeCurrentStep ()
{
    struct timespec ts1                 = {0};
    struct timespec ts2                 = {0};
    struct timespec tsr1                = {0};
    struct timespec tsr2                = {0};
            SI32    returnValue         = -1;
            SI32    returnValueR        = -1;
            UI32    operationCode       = 0;
            UI32    numberOfSeconds     = 0;
            SI32    numberOfNanoSeconds = 0;
            UI32    currentStep         = m_currentStep;
            UI32    numberOfSteps       = m_numberOfSteps;

    if (currentStep < (numberOfSteps - 2))
    {
        if ((0 != m_clockId) && (NULL != m_pWaveMessage) && (NULL != m_pWaveElement))
        {
            operationCode = m_pWaveMessage->getOperationCode ();

            returnValue   = clock_gettime (m_clockId,      &ts1);
            returnValueR  = clock_gettime (CLOCK_REALTIME, &tsr1);
        }
    }


    ResourceId status = (m_pWaveElement->*(m_pSteps[currentStep])) (this);

    if (currentStep < (numberOfSteps - 2))
    {
        if (0 == returnValue)
        {
            returnValue = clock_gettime (m_clockId, &ts2);

            if (0 == returnValue)
            {
                numberOfSeconds     = ts2.tv_sec  - ts1.tv_sec;
                numberOfNanoSeconds = ts2.tv_nsec - ts1.tv_nsec;

                m_pWaveElement->updateTimeConsumedInThisThread (operationCode, currentStep, numberOfSeconds, numberOfNanoSeconds);
            }
        }

        if (0 == returnValueR)
        {
            returnValueR = clock_gettime (CLOCK_REALTIME, &tsr2);

            if (0 == returnValueR)
            {
                numberOfSeconds     = tsr2.tv_sec  - tsr1.tv_sec;
                numberOfNanoSeconds = tsr2.tv_nsec - tsr1.tv_nsec;

                m_pWaveElement->updateRealTimeConsumedInThisThread (operationCode, currentStep, numberOfSeconds, numberOfNanoSeconds);
            }
        }
    }

    return (status);
}

ResourceId WaveSynchronousLinearSequencerContext::getCompletionStatus ()
{
    return (m_completionStatus);
}

ResourceId WaveSynchronousLinearSequencerContext::execute ()
{
    UI32 i;
    UI32 effectiveNumberOfSteps = m_numberOfSteps - 2;
    SI32 returnValue            = 0;

    m_completionStatus = WAVE_MESSAGE_ERROR;

    returnValue = pthread_getcpuclockid (pthread_self (), &m_clockId);

    if (0 != returnValue)
    {
        m_clockId = 0;

        cerr << "Could not get Clock ID for this thread." << endl;
    }

    for (i = 0; i < effectiveNumberOfSteps; i++)
    {
        m_completionStatus = executeCurrentStep ();

        if (WAVE_MESSAGE_SUCCESS != m_completionStatus)
        {
            if (WAVE_SEQUENCER_SKIP_TO_SUCCESS_STEP == m_completionStatus)
            {
                // Move step pointer to the success step
                m_currentStep = effectiveNumberOfSteps;
                m_completionStatus = WAVE_MESSAGE_SUCCESS;
            }

            break;
        }
        else
        {
            advanceCurrentStep ();
        }
    }

    if (WAVE_MESSAGE_SUCCESS != m_completionStatus)
    {
        m_currentStep = m_numberOfSteps - 1;
    }

    return (executeCurrentStep ());
}

WaveMessage *WaveSynchronousLinearSequencerContext::getPWaveMessage ()
{
    return (m_pWaveMessage);
}

void WaveSynchronousLinearSequencerContext::setPWaveMessage (WaveMessage *pWaveMessage)
{
    m_pWaveMessage = pWaveMessage;
}

WaveAsynchronousContext *WaveSynchronousLinearSequencerContext::getPWaveAsynchronousContext ()
{
    return (m_pWaveAsynchronousContext);
}

bool WaveSynchronousLinearSequencerContext::getIsTransactionStartedByMe () const
{
    return (m_isTransactionStartedByMe);
}

void WaveSynchronousLinearSequencerContext::setIsTransactionStartedByMe (const bool &isTransactionStartedByMe)
{
    m_isTransactionStartedByMe = isTransactionStartedByMe;
}

void WaveSynchronousLinearSequencerContext::addManagedObjectsForGarbageCollection (const vector<WaveManagedObject *> &managedObjectsForGarbageCollection)
{
    m_managedObjectsForGarbageCollection.insert (m_managedObjectsForGarbageCollection.end (), managedObjectsForGarbageCollection.begin (), managedObjectsForGarbageCollection.end ());
}

void WaveSynchronousLinearSequencerContext::addManagedObjectForGarbageCollection (WaveManagedObject *pWaveManagedObjectForGarbageCollection)
{
    waveAssert (NULL != pWaveManagedObjectForGarbageCollection, __FILE__, __LINE__);

    m_managedObjectsForGarbageCollection.push_back (pWaveManagedObjectForGarbageCollection);
}

void WaveSynchronousLinearSequencerContext::garbageCollect ()
{
    UI32 numberOfManagedObjectsForGarbageCollection = m_managedObjectsForGarbageCollection.size ();
    UI32 i                                          = 0;

    for (i = 0; i < numberOfManagedObjectsForGarbageCollection; i++)
    {
        waveAssert (NULL != m_managedObjectsForGarbageCollection[i], __FILE__, __LINE__);

        delete (m_managedObjectsForGarbageCollection[i]);
    }
}

bool WaveSynchronousLinearSequencerContext::getIsADelayedCommitTransaction () const
{
    return (m_isADelayedCommitTransaction);
}
 
void WaveSynchronousLinearSequencerContext::setIsADelayedCommitTransaction (const bool &isADelayedCommitTransaction)
{
    m_isADelayedCommitTransaction = isADelayedCommitTransaction;
}

}

File: Java/source/com/CxWave/Wave/Framework/Utils/Sequencer/WaveSynchronousLinearSequencerContext.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Utils.Sequencer;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Vector;

import com.CxWave.Wave.Framework.Messaging.Local.WaveMessage;
import com.CxWave.Wave.Framework.ObjectModel.WaveElement;
import com.CxWave.Wave.Framework.ObjectModel.WaveManagedObject;
import com.CxWave.Wave.Framework.Type.UI32;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Framework.Utils.Context.WaveAsynchronousContext;
import com.CxWave.Wave.Framework.Utils.Source.WaveJavaSourceRepository;
import com.CxWave.Wave.Framework.Utils.Time.StopWatch;
import com.CxWave.Wave.Framework.Utils.Time.ThreadStopWatch;
import com.CxWave.Wave.Framework.Utils.Trace.WaveTraceUtils;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;

public class WaveSynchronousLinearSequencerContext
{
    WaveMessage                             m_waveMessage                 = null;
    WaveAsynchronousContext                 m_waveAsynchronousContext     = null;
    private WaveElement                     m_waveElement                 = null;
    private String[]                        m_steps                       = null;
    private int                             m_numberOfSteps               = 0;
    private int                             m_currentStep                 = 0;
    private ResourceId                      m_completionStatus            = ResourceId.WAVE_MESSAGE_ERROR;
    private boolean                         m_isTransactionStartedByMe    = false;

    private final Vector<WaveManagedObject> m_queryResults                = new Vector<WaveManagedObject> ();

    boolean                                 m_isADelayedCommitTransaction = false;

    private final StopWatch                 m_wallClockTimeStopWatch      = new StopWatch ();
    private final ThreadStopWatch           m_threadTimeStopWatch         = new ThreadStopWatch ();

    private final UI32                      m_operationCode               = new UI32 (0);

    private final Vector<Method>            m_methodsForSteps             = new Vector<Method> ();

    public WaveSynchronousLinearSequencerContext (final WaveMessage waveMessage, final WaveElement waveElement, final String[] steps)
    {
        m_waveMessage = waveMessage;
        m_waveElement = waveElement;
        m_steps = steps;
        m_numberOfSteps = steps.length;

        final boolean isValid = validateAndCompute ();

        if (!isValid)
        {
            WaveAssertUtils.waveAssert ();
        }
    }

    public WaveSynchronousLinearSequencerContext (final WaveAsynchronousContext waveAsynchronousContext, final WaveElement waveElement, final String[] steps)
    {
        m_waveAsynchronousContext = waveAsynchronousContext;
        m_waveElement = waveElement;
        m_steps = steps;
        m_numberOfSteps = steps.length;

        final boolean isValid = validateAndCompute ();

        if (!isValid)
        {
            WaveAssertUtils.waveAssert ();
        }
    }

    private boolean validateAndCompute ()
    {
        if (null == m_waveElement)
        {
            return (false);
        }

        if (null == m_steps)
        {
            return (false);
        }

        if (3 > m_numberOfSteps)
        {
            WaveTraceUtils.fatalTracePrintf ("WaveSynchronousLinearSequencerContext.validateAndCompute :  There should be atleast three steps to run a Wave Linear Sequencer. %s Steps were specified.", m_numberOfSteps);

            return (false);
        }

        final Class<?> waveElementClass = m_waveElement.getClass ();

        WaveAssertUtils.waveAssert (null != waveElementClass);

        for (final String methodNameForStep : m_steps)
        {
            WaveTraceUtils.infoTracePrintf ("WaveSynchronousLinearSequencerContext.validateAndCompute : Searching for method name : %s", methodNameForStep);

            final Class<?> classToSearchForTheMethod = waveElementClass;
            Method methodForStep = null;

            methodForStep = WaveJavaSourceRepository.getMethodForWaveSynchronousLinearSequencerStepInWaveJavaClass (classToSearchForTheMethod.getTypeName (), methodNameForStep);

            if (null != methodForStep)
            {
                m_methodsForSteps.add (methodForStep);
            }
            else
            {
                WaveTraceUtils.fatalTracePrintf ("WaveSynchronousLinearSequencerContext.validateAndCompute : Could not obtain the Step method %s on class %s", methodNameForStep, waveElementClass.getName ());
                WaveAssertUtils.waveAssert ();

                return (false);
            }
        }

        return (true);
    }

    private void advanceCurrentStep ()
    {
        m_currentStep++;
    }

    public ResourceId getCompletionStatus ()
    {
        return (m_completionStatus);
    }

    public ResourceId executeCurrentStep ()
    {
        if (m_currentStep < (m_numberOfSteps - 2))
        {
            if (null != m_waveMessage)
            {
                m_operationCode.setValue (m_waveMessage.getOperationCode ());
            }

            m_wallClockTimeStopWatch.start ();
            m_threadTimeStopWatch.start ();
        }

        final Method methodForCurrentStep = m_methodsForSteps.get (m_currentStep);

        WaveAssertUtils.waveAssert (null != methodForCurrentStep);

        WaveAssertUtils.waveAssert (null != m_waveElement);

        Object returnValueObject = null;

        try
        {
            returnValueObject = methodForCurrentStep.invoke (m_waveElement, this);
        }
        catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e)
        {
            WaveTraceUtils.fatalTracePrintf ("WaveSynchronousLinearSequencerContext.executeCurrentStep : Failed to invoke method : %s, Details : %s", methodForCurrentStep.getName (), e.toString ());
        }

        final ResourceId status = (ResourceId) returnValueObject;

        WaveAssertUtils.waveAssert (null != status);

        if (m_currentStep < (m_numberOfSteps - 2))
        {
            m_wallClockTimeStopWatch.stop ();
            m_threadTimeStopWatch.stop ();

            m_waveElement.updateTimeConsumedInThisThread (m_operationCode, m_currentStep, m_threadTimeStopWatch.getLastLapDuration ());
            m_waveElement.updateRealTimeConsumedInThisThread (m_operationCode, m_currentStep, m_wallClockTimeStopWatch.getLastLapDuration ());
        }

        return (status);
    }

    public ResourceId execute ()
    {
        int i;
        final int effectiveNumberOfSteps = m_numberOfSteps - 2;

        m_completionStatus = ResourceId.WAVE_MESSAGE_ERROR;

        for (i = 0; i < effectiveNumberOfSteps; i++)
        {
            m_completionStatus = executeCurrentStep ();

            if (!(ResourceId.WAVE_MESSAGE_SUCCESS.equals (m_completionStatus)))
            {
                if (ResourceId.WAVE_SEQUENCER_SKIP_TO_SUCCESS_STEP.equals (m_completionStatus))
                {
                    // Move step pointer to the success step
                    m_currentStep = effectiveNumberOfSteps;
                    m_completionStatus = ResourceId.WAVE_MESSAGE_SUCCESS;
                }

                break;
            }
            else
            {
                advanceCurrentStep ();
            }
        }

        if (!(ResourceId.WAVE_MESSAGE_SUCCESS.equals (m_completionStatus)))
        {
            m_currentStep = m_numberOfSteps - 1;
        }

        return (executeCurrentStep ());
    }

    public WaveMessage getWaveMessage ()
    {
        return m_waveMessage;
    }

    public void setWaveMessage (final WaveMessage waveMessage)
    {
        m_waveMessage = waveMessage;
    }

    public WaveAsynchronousContext getWaveAsynchronousContext ()
    {
        return m_waveAsynchronousContext;
    }

    public void setWaveAsynchronousContext (final WaveAsynchronousContext waveAsynchronousContext)
    {
        m_waveAsynchronousContext = waveAsynchronousContext;
    }

    public boolean getIsADelayedCommitTransaction ()
    {
        return m_isADelayedCommitTransaction;
    }

    public void setIsADelayedCommitTransaction (final boolean isADelayedCommitTransaction)
    {
        m_isADelayedCommitTransaction = isADelayedCommitTransaction;
    }

    public boolean getIsTransactionStartedByMe ()
    {
        return m_isTransactionStartedByMe;
    }

    public void setIsTransactionStartedByMe (final boolean isTransactionStartedByMe)
    {
        m_isTransactionStartedByMe = isTransactionStartedByMe;
    }
}
